test/model/message_test.dart

```diff
@@ -123,6 +127,179 @@ void main() {
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('request fails', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+    }));
+
+    test('request fails; take failed edit', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(apiException: eg.apiBadRequest(), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+
+      check(store.takeFailedMessageEdit(message.id)).equals('new content');
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('takeFailedMessageEdit throws StateError when nothing to take', () => awaitFakeAsync((async) async {
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+      check(() => store.takeFailedMessageEdit(message.id)).throws<StateError>();
+    }));
+
+    test('request failure after event arrival', () => awaitFakeAsync((async) async {
+      // This can happen with network issues.
+
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(
+        httpException: const SocketException('failed'), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+
+      async.elapse(Duration(milliseconds: 500));
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      async.elapse(Duration(milliseconds: 500));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('request failure before event arrival', () => awaitFakeAsync((async) async {
+      // This can happen with network issues.
+
+      await prepare();
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+
+      connection.prepare(
+        httpException: const SocketException('failed'), delay: Duration(seconds: 1));
+      store.editMessage(messageId: message.id, content: 'new content');
+
+      async.elapse(Duration(seconds: 1));
+      check(store.getEditMessageErrorStatus(message.id)).isNotNull().isTrue();
+
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      check(store.getEditMessageErrorStatus(message.id)).isNull();
+    }));
+
+    test('request failure before event arrival; take failed edit in between', () => awaitFakeAsync((async) async {
```

Comment:
nit: the format "request fails, then message deleted" seems preferable to me to help understand the sequence of events